harness/determined/keras/_tf_keras_trial.py

```diff
@@ -26,49 +29,52 @@
 IMPOSSIBLY_LARGE_EPOCHS = sys.maxsize
 
 
-def load_optimizer_weights(model: Model, load_path: pathlib.Path) -> None:
+def load_optimizer_weights(
+    model: Model, h5group: Any, optimizer: Optional[tf.keras.optimizers.Optimizer]
+) -> None:
     """
     Load the optimizer states from a tf.keras model saved with
     tf.keras.models.save_model(). Ignores and prints a warning message when
     encountering a graph network. This implementation is lifted from
     tf.keras.models.load_model().
     """
-    f = h5py.File(str(load_path), mode="r")
-    if "optimizer_weights" in f:
-        tf2_2_or_newer = version.parse(tf.__version__) >= version.parse("2.2.0")
-        if model._is_graph_network or tf2_2_or_newer:  # pylint: disable=protected-access
-            if tf2_2_or_newer:
-                try:
-                    model.optimizer._create_all_weights(model.trainable_variables)
-                except (NotImplementedError, AttributeError):
-                    logging.warning(
-                        "Error when creating the weights of optimizer, making it "
-                        "impossible to restore the saved optimizer state. As a result, "
-                        "your model is starting with a freshly initialized optimizer."
-                    )
-            else:
-                # Build train function (to get weight updates).  Models that aren't
-                # graph networks must wait until they are called with data to
-                # _make_train_function() and so can't load optimizer weights.
-                model._make_train_function()
+    tf2_2_or_newer = version.parse(tf.__version__) >= version.parse("2.2.0")
+    if model._is_graph_network or tf2_2_or_newer:  # pylint: disable=protected-access
+        if not optimizer:
+            optimizer = model.optimizer
 
-            optimizer_weight_values = load_optimizer_weights_from_hdf5_group(f)
+        if tf2_2_or_newer:
             try:
-                model.optimizer.set_weights(optimizer_weight_values)
-            except ValueError:
+                optimizer._create_all_weights(model.trainable_variables)
```

Comment:
Might be worth a comment explaining what's going on here.  I find it pretty hard to understand the control flow here

Reply:
honestly that's probably why it is so hard to understand.